pkg/cron/locker.go

```diff
@@ -119,136 +123,117 @@ func (l *Locker) createLock(ctx context.Context, key string) error {
-					current.ExpireAt < time.Now().Unix() {
-					// Should not renew.
-					return nil
-				}
-
-				// Renew key.
-				expiry := time.Now().Unix() + int64(l.expiry.Seconds())
-
-				_, err = tx.DistributeLock.Update().
-					Where(distributelock.ID(key)).
-					SetExpireAt(expiry).
-					Save(ctx)
-				if err != nil && !model.IsNotFound(err) {
-					return err
-				}
-
-				l.logger.V(6).Infof("success renew key %s", key)
-
-				return nil
-			})
-			if err != nil {
-				l.logger.Warnf("error renew key %s: %v", key, err)
-			}
-		case <-ctx.Done():
-			return
-		}
+	if err != nil {
+		return nil, err
 	}
+
+	return &Lock{
+		logger: l.logger.WithName("locker").
+			WithValues("key", key, "id", strs.Hex(6)),
+		holder: l.holder,
+		client: l.client,
+		expiry: l.expiry,
+		key:    key,
+	}, nil
 }
 
 // Lock implement the cronjob go-co-op/gocron Lock interface.
 type Lock struct {
-	key         string
-	client      model.ClientSet
-	logger      log.Logger
-	cancelRenew context.CancelFunc
+	logger log.Logger
+	holder string
+	client model.ClientSet
+	expiry time.Duration
+	key    string
 }
 
-// Unlock release the locked key.
-func (l *Lock) Unlock(ctx context.Context) (err error) {
-	l.logger.V(6).Infof("try to unlock key %s", l.key)
-
-	defer func() {
+// Renew keeps alive the lock, returns true if not found,
+// and never stop until context canceled.
+func (lk *Lock) Renew(ctx context.Context) (done bool, err error) {
+	err = lk.client.WithTx(ctx, func(tx *model.Tx) error {
+		// Get key.
+		key, err := tx.DistributeLock.Query().
+			Where(
+				distributelock.ID(lk.key),
+				distributelock.Holder(lk.holder)).
+			Only(ctx)
 		if err != nil {
-			l.logger.Warnf("error unlock key %s: %v", l.key, err)
+			return err
 		}
-		// Cancel renew goroutine.
-		l.cancelRenew()
-	}()
-
-	err = l.deleteLock(ctx, l.key)
-	if err != nil {
-		return err
-	}
-
-	l.logger.V(6).Infof("success unlock key %s", l.key)
 
-	return nil
-}
-
-// deleteLock delete the lock with key.
-func (l *Lock) deleteLock(ctx context.Context, key string) error {
-	return l.client.WithTx(ctx, func(tx *model.Tx) (err error) {
-		// Lock key for delete.
-		current, err := tx.DistributeLock.Query().
-			Where(distributelock.ID(key)).
-			ForUpdate().
-			Only(ctx)
-		if err != nil && !model.IsNotFound(err) {
-			return err
+		if key.ExpireAt < time.Now().Unix() {
+			// Should not renew.
+			return nil
 		}
 
-		// Delete key.
-		err = tx.DistributeLock.DeleteOne(current).
+		// Renew key.
+		expiry := time.Now().Unix() + int64(lk.expiry.Seconds())
+
+		err = tx.DistributeLock.UpdateOne(key).
+			SetExpireAt(expiry).
 			Exec(ctx)
 		if err != nil {
 			return err
 		}
 
-		return
+		lk.logger.V(6).Info("succeeded renew")
+
+		return nil
 	})
+	if err != nil && ctx.Err() == nil {
+		if model.IsNotFound(err) {
+			// Not found, should not renew again.
+			return true, nil
+		}
+
+		lk.logger.Warnf("error renewing, try at next loop: %v", err)
+	}
+
```

Comment:
while the job is finished, and the key is unlocked, seems like the renew goroutine is still here, while the current instance still successfully gets lock next time, the renew goroutine can't exit